File: Lab7/ExpressionCalcStatic/ExpressionCalc.cpp
#include "ExpressionCalc.h"
#include "ManualSQRT.h"
#include <format>
#include <iostream>

using namespace std;

    double ExpressionCalc::GetResult()
    {
        return result;
    }    
    double ExpressionCalc::CalculateExpression()
    {
        double numerator = a * b / 4 - 1;
        double innerExpression = 41 - d;
        ManualSQRT sqrt(innerExpression);
        double denominator = sqrt.GetNumber() - b * a + c;

        if (denominator == 0)
        {
            throw runtime_error("Denominator cannot be zero");
        }
        else
        {
            return numerator / denominator;
        }
    }
    ExpressionCalc::ExpressionCalc(double a1, double b1, double c1, double d1)
    {
        a = a1;
        b = b1;
        c = c1;
        d = d1;

        try
        {
            result = CalculateExpression();
        }
        catch (const runtime_error& e)
        {
            cout << ("Error: ") << e.what() << endl;
        }
    }

File: Lab7/ExpressionCalcDLL/ExpressionCalc.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace ExpressionCalcDLL
{
    public class ExpressionCalc
    {
        private double a,b,c,d;
        private double result;

        public double GetResult()
        { return result; } 
        public ExpressionCalc(double a, double b, double c, double d)
        {
            this.a = a;
            this.b = b;
            this.c = c;
            this.d = d;

            try
            {
                result = this.CalculateExpression();
            }
            catch (ArgumentException exception)
            {
                Console.WriteLine("Error: " + exception.Message);
            }
        }

        public double CalculateExpression()
        {
            double numerator = a * b / 4 - 1;
            double innerExpression = 41 - d;
            ManualSQRT sqrt = new ManualSQRT(innerExpression);
            double denominator = sqrt.GetNumber() - b*a + c;

            if (denominator == 0)
            {
                throw new ArgumentException("Denominator cannot be zero");
            }
            else
            {
                return numerator / denominator;
            }
        }
    }
}
